packages/passwordless-auth-construct/src/magic-link/construct.ts

```diff
@@ -0,0 +1,135 @@
+import { Construct } from 'constructs';
+import { CustomAuthTriggers, MagicLinkAuthOptions } from '../types.js';
+import { Aws, RemovalPolicy, Stack } from 'aws-cdk-lib';
+import { IKey, Key, KeySpec, KeyUsage } from 'aws-cdk-lib/aws-kms';
+import {
+  AccountRootPrincipal,
+  Effect,
+  PolicyDocument,
+  PolicyStatement,
+} from 'aws-cdk-lib/aws-iam';
+import { AttributeType, BillingMode, Table } from 'aws-cdk-lib/aws-dynamodb';
+
+/**
+ * Amplify Auth CDK Construct
+ */
+export class AmplifyMagicLinkAuth extends Construct {
+  /**
+   * Create a new Auth construct with AuthProps.
+   * If no props are provided, email login and defaults will be used.
+   */
+  constructor(
+    scope: Construct,
+    id: string,
+    triggers: CustomAuthTriggers,
+    props: MagicLinkAuthOptions
+  ) {
+    super(scope, id);
+
+    if (!props) {
+      return;
+    }
+
+    const kmsKey = this.createKmsKey(scope, id, triggers);
+    const secretsTable = this.createSecretsTable(scope, id, triggers);
+
+    const createAuthChallengeEnvVars = {
+      magicLinkFromAddress:
+        typeof props.email === 'boolean' ? undefined : props.email?.fromAddress,
+      magicLinkAllowedOrigins: props.allowedOrigins.join(','),
+      magicLinkKmsKeyId: kmsKey.keyId,
+      magicLinkTableName: secretsTable.tableName,
+      magicLinkSecondsUntilExpiry:
+        props.linkDuration?.toSeconds().toString() ?? '900',
+    };
+
+    const verifyAuthChallengeResponseEnvVars = {
+      magicLinkTableName: secretsTable.tableName,
+    };
+
+    for (const [key, value] of Object.entries(createAuthChallengeEnvVars)) {
+      triggers.createAuthChallenge.addEnvironment(key, value ?? '');
+    }
+
+    for (const [key, value] of Object.entries(
+      verifyAuthChallengeResponseEnvVars
+    )) {
+      triggers.verifyAuthChallengeResponse.addEnvironment(key, value);
+    }
+  }
+
+  private createKmsKey = (
+    scope: Construct,
+    id: string,
+    triggers: CustomAuthTriggers
+  ): IKey => {
+    const alias = `${id}-${Stack.of(scope).stackName}`;
+    const key = new Key(this, `KmsKeyRsa${id}`, {
+      keySpec: KeySpec.RSA_2048,
+      keyUsage: KeyUsage.SIGN_VERIFY,
+      alias: alias,
+      removalPolicy: RemovalPolicy.DESTROY,
+      policy: new PolicyDocument({
+        statements: [
+          new PolicyStatement({
+            effect: Effect.ALLOW,
+            notActions: ['kms:Sign'],
+            resources: ['*'],
+            principals: [new AccountRootPrincipal()],
+          }),
+        ],
+      }),
+    });
+
+    const permissions = {
+      effect: Effect.ALLOW,
+      resources: [
+        `arn:${Aws.PARTITION}:kms:${Aws.REGION}:${Aws.ACCOUNT_ID}:key/*`,
+      ],
+      actions: ['kms:Sign'],
+    };
+    key.addToResourcePolicy(
+      new PolicyStatement({
+        ...permissions,
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
please either add comment explaining why this is ok or `if` statement to handle it (and throw some meaningfull error).